### ISSUE TITLE
AddFeat: I implemented the features mentioned in this paper(SpecAugment) in chainbin and i want to know what tests do I need to do?

### ISSUE THREAD

**Issue by bringtree** [kaldi-asr/kaldi#3291]
SpecAugment: https://arxiv.org/pdf/1904.08779.pdf
Hello, I implemented the features mentioned in this paper(SpecAugment) in chainbin and i want to know what tests do I need to do?

**Comment by danpovey** [comment]
I don't know what you are saying.

**Comment by danpovey** [comment]
... I suppose you could just run one of the training scripts, e.g. in mini_librispeech, but I am confused, as if you didn't already know how to do this, I don't see that you would be able to implement SpecAugment.

**Comment by harry0731** [comment]
I'm interesting to try on this new feature. Can you share how to implement it?

**Comment by bringtree** [comment]
@harry0731   Sorry, the code belongs to my company. If my leader allowed me to share the source code. I would like to do it.  
Now I am testing the model. If the code good for the model, I will request my leader to share the source code.

**Comment by bringtree** [comment]
@harry0731 Sorry,   I can't get a score better than the baseline when I use SpecAugment.
I had done many experiments.  I feel despair.

**Comment by danpovey** [comment]
In our experiments we have not got a very large improvement, only about 2%
relative IIRC (Phani Sankar was doing these experiments).  You might need
to tune it a bit.  I think the dropped frequency ranges are the most
important.  Remember to replace each filterbank coefficient with the mean
value within that utterance (since our features do not have CMN applied).

**Comment by bringtree** [comment]
I always use mfcc feature. I am not sure these experiments would get an improvement on using mfcc feature.

**Comment by danpovey** [comment]
You should probably project the MFCC features back to log energies before
doing the SpecAugment stuff.
If you look for our nnet3 examples with `cnn` in them, e.g. look for
run_cnn_tdnn.sh, you will see an example of projecting MFCC to log-mels.
 We normally write as MFCC because it's easier to compress, being more
decorrelated.

Dan

**Comment by bringtree** [comment]
I found the reason why I can't get a score better than the baseline.
After I masked the feature. I do the batch normalization in the next layer. 

Understanding the Disharmony between Dropout and Batch Normalization by Variance Shift.

I think "specArugment" like "dropout". So I needn't do the batch normalization in the next layer.

**Comment by danpovey** [comment]
Hm, it might cause a mismatch between training and test if you put
batchnorm after SpecAugment.
IIRC I put batchnorm *before* SpecAugment in the stuff we checked in.
